### DISCUSSION TITLE
Railsにデフォルトで存在しないレイヤー(servicesなど）なにか導入してますか？

### DISCUSSION THREAD

**Discussion by danimal141** [roppongirb/discussions#32] [Ideas]

**Comment by hatappi** [comment]
## 概要
Rails way にのると model, controller, view を使ってアプリケーションを構築していくけどよく service layer いれたりするけどどうなん？

**Comment by hatappi** [comment]
## case1
https://github.com/okuramasafumi/alba を導入した時に Serializer directory ができた。

gem 起因で作成される

**Reply by hatappi** [reply]
decorater 関連の gem とかもこのパターン

**Reply by hatappi** [reply]
view component とかもそう
https://github.com/github/view_component

**Comment by hatappi** [comment]
## case2: Adapters を作る
Twitter library のような 3rd party を使うためのラッパー目的の層

**Reply by hatappi** [reply]
## Other cases
- app/lib におく
   - services direcotry を生やしてそこにおく
   - 内部のサービスのリクエストもここにおく
- model に namespace きっておく

  ```rb
  module Twitter
    class Client
    end
  end
  ```

**Reply by hatappi** [reply]
## app/lib におく辛み

実装者の判断に委ねられるので model にあったり、lib にあったりみたいなことが起きる？

**Comment by hatappi** [comment]
## Rails に存在しないレイヤーをおいた時のつらみ
実装者の判断によってしまうので、実装が重複してしまうことも。。。

**Reply by scrpgil** [reply]
libにもmodelsにもcontrollersにも同じ名前のファイルができるんですよねぇ。。。

**Comment by hatappi** [comment]
## Rails 公式はなんていってる？
Fat model?

**Reply by hatappi** [reply]
Scope が増えすぎた時は module に切り出して対応

**Comment by hatappi** [comment]
Perfect Ruby on Rails だと Service object 作るみたいな例がある。

Service Object に具体的な名前 (`trasfer_xxx`) をつけることでなんでもいれる Service object にしない

**Reply by hatappi** [reply]
チームでどういう単位で区切るとかを話ておくのが大事

**Reply by scrpgil** [reply]
「#{モデル名}Service」みたいなサービスオブジェクト作ると、やっぱり曖昧になりますよねぇ。

**Reply by nyamadori** [reply]
パーフェクト Ruby on Rails の「イベントの見落としがないかを確認する」もいい話でした。（Service を置かずにモデルだけで実装できる方法）。
